File: problems/baires/owo3.js
function foo(s1, s2) {
	let diff = 0;
	for (let i = 0; i < s1.length; ++i) {
		if (s1[i] != s2[i]) ++diff;
		if (diff > 1) return false;
	}
	return true;
}

function bfs(graph, size, origin) {
	const q = []; // queue
	const dis = []; // dis[x] = minimum distance from origin to x
	const vis = []; // visited
	for	(let i = 0; i < size; ++i) {
		dis[i] = 700;
		vis[i] = false;
	}
	dis[origin] = 0;
	vis[origin]=true;
	q.push(origin);
	while (q.length > 0) {
		const x = q[0];
		q.shift();
		graph[x].forEach(v => {
			if(!vis[v]){
				dis[v]=dis[x]+1;
				vis[v]=true;
				q.push(v);
			}
		})
	}
	return dis;
}

function wordLadder(beginWord, endWord, wordList) {
	const list = [beginWord, endWord, ...wordList];
	const size = list.length;
	const graph = [];
	for (let i = 0; i < size; ++i) {
		graph[i] = [];
		for (let j = 0; j < size; ++j) {
			if(i !== j && foo(list[i], list[j])){
				graph[i].push(j);
			}
		}
	}
	const dis = bfs(graph, size, 0);
	const ans = dis[1] + 1;
	return ans != 700 ? ans : 0;
}

console.log(wordLadder("hit", "cog", ["hot", "dot", "dog", "lot", "log", "cog"]));

File: problems/baires/owo3.py
def dist(s1, s2):
    if len(s1) > len(s2):
        s1, s2 = s2, s1

    distances = range(len(s1) + 1)
    for i2, c2 in enumerate(s2):
        distances_ = [i2+1]
        for i1, c1 in enumerate(s1):
            if c1 == c2:
                distances_.append(distances[i1])
            else:
                distances_.append(1 + min((distances[i1], distances[i1 + 1], distances_[-1])))
        distances = distances_
    return distances[-1]


def search(s1, s2, wordList, h):
    if s1 == s2:
        return h
    if not wordList:
        return 700
    ans = 700
    for s in wordList:
        d = dist(s1, s)
        if d == 1:
            wordList2 = [x for x in wordList if s != x]
            ans = min(ans, search(s, s2, wordList2, h + 1))
    return ans


def wordLadder(beginWord, endWord, wordList):
    ans = search(beginWord, endWord, wordList, 1)
    return ans if ans != 700 else 0

#print(wordLadder("hit", "cog", ["hot", "dot", "dog", "lot", "log", "cog"]))

